proxy/create_container_interceptor.go

```diff
@@ -76,48 +83,114 @@ func (i *createContainerInterceptor) InterceptRequest(r *http.Request) error {
 	return nil
 }
 
-func (i *createContainerInterceptor) addWeaveWaitVolume(hostConfig *docker.HostConfig) {
+func (i *createContainerInterceptor) containerHostname(r *http.Request, container map[string]interface{}) (hostname string, err error) {
+	hostname = r.URL.Query().Get("name")
+	if i.proxy.Config.HostnameFromLabel != "" {
+		hostname, err = i.hostnameFromLabel(hostname, container)
+	}
+	hostname = i.proxy.hostnameMatchRegexp.ReplaceAllString(hostname, i.proxy.HostnameReplacement)
+	return
+}
+
+func (i *createContainerInterceptor) hostnameFromLabel(hostname string, container map[string]interface{}) (string, error) {
+	labels, err := lookupObject(container, "Labels")
+	if err != nil {
+		return "", err
+	}
+
+	labelIface, ok := labels[i.proxy.Config.HostnameFromLabel]
+	if !ok || labelIface == nil {
+		return hostname, nil
+	}
+
+	label, ok := labelIface.(string)
+	if !ok {
+		return "", &UnmarshalWrongTypeError{i.proxy.Config.HostnameFromLabel, "string", labelIface}
+	}
+
+	return label, nil
+}
+
+func (i *createContainerInterceptor) addWeaveWaitVolume(hostConfig map[string]interface{}) error {
+	configBinds, err := lookupStringArray(hostConfig, "Binds")
+	if err != nil {
+		return err
+	}
+
 	var binds []string
-	for _, bind := range hostConfig.Binds {
+	for _, bind := range configBinds {
 		s := strings.Split(bind, ":")
 		if len(s) >= 2 && s[1] == "/w" {
 			continue
 		}
 		binds = append(binds, bind)
 	}
-	hostConfig.Binds = append(binds, fmt.Sprintf("%s:/w:ro", i.proxy.weaveWaitVolume))
+	hostConfig["Binds"] = append(binds, fmt.Sprintf("%s:/w:ro", i.proxy.weaveWaitVolume))
+	return nil
 }
 
-func (i *createContainerInterceptor) setWeaveWaitEntrypoint(container *docker.Config) error {
-	if len(container.Entrypoint) == 0 {
-		image, err := i.proxy.client.InspectImage(container.Image)
+func (i *createContainerInterceptor) setWeaveWaitEntrypoint(container map[string]interface{}) error {
+	var entrypoint []string
+	if e, ok := container["Entrypoint"]; ok && e != nil {
+		switch e := e.(type) {
+		case string:
+			entrypoint = []string{e}
+		case []string:
```

Comment:
It looks like the rest of this block duplicates lookupStringArray(); could just call it?